### DISCUSSION TITLE
Next Version Features

### DISCUSSION THREAD

**Discussion by mattsta** [mattsta/icli#15] [General]
Just a heads up for some things I have queued up to release but haven't finalized yet:

- color themes can be changed at runtime (`colorset` command, changes are remembered across restarts)
- automatic _correct_ price rounding for _all_ instruments using IBKR's market price rule APIs
  - e.g. equities trade in $0.01 increments, but options can be $0.01 or $0.05 or $0.10 or $0.25 increments, while some futures are $0.001 or $0.002 or $0.25 or $5 or 0.00390625 or 0.03125 (and more!) increments, etc.
  - have you ever had an order rejected because the price was "wrong?" well, never again!
  - now we use IBKR APIs to fetch (and _cache_ locally) the _correct price math_ for all instruments across all price sizes and we enforce correct rounding at all order placement or order modification points.
- easy straddle/strangle creation and vertical call/put creation all combined in one command
  - `straddle /ES 20` would add a quote for a straddle (long call, long put) ±20 points from live /ES ATM
  - `straddle /ES 0` would give you an ATM strangle
  - `straddle /ES vertical put -10 -20` would give you a vertical put with the long leg -10 points from ATM and the short leg -20 points from ATM (e.g. /ES 5000 with vertical put -10 -20 would add a quote for BUY PUT 4980, SELL PUT 4960)
    - also, IBKR treats quotes as long by default, but if you sell a long quote, IBKR automatically flips the sides, so you can sell long quotes as a short spread. e.g. if you SELL (BUY PUT 4980, SELL PUT 4960) it actually generates SELL PUT 4980, BUY PUT 4960.
  - `straddle /ES vertical call 10 20` creates a spread quote for buy call +10 ATM and sell call +20 ATM.
  - You can also combine the syntax in any combination (also with shorthand): `straddle /ES v c 10 20 v p -10 -20` would create a quote for (BUY CALL +10 ATM, SELL CALL +20 ATM, BUY PUT -10 ATM, SELL PUT -20 ATM) all as a single spread.
  - by default, it uses the closest contract expiration, but you can alter how far the automatic quotes expire with a global setting of `set dte N` where `N` is days away from today. `set dte` can also accept a day name as a word (`set dte thursday`) and it will figure out how many days away to use for `N`
- `buy` command input is now a full parser ([`orderlang`](https://github.com/mattsta/tradeapis/blob/main/tradeapis/orderlang.py) in tradeapis) instead of just 3 command fields — we have much more flexibility for parameter order, types of information (attached take-profit, stop-loss either one or both and optional algo overrides for each attached order), and config options now (trailing stop orders with initial stop, trail, limit prices all in the `buy` command).
- `buy` command is improved to use live event updates instead of sleeping timers for detecting when to try and move a limit price closer to market for more aggressive fills, so it should be much more responsive. Also it has some extra guards to try and protect against adversarial quotes.
- `buy` command can also generate stop limit and trailing stop limit orders for brackets too.
- `buy` command can handle all instrument types including spreads/bags both long and short. IBKR allows placing bracket orders (or single take-profit, stop-loss) on spreads too, so all parameter combinations work for all instrument types.
- all price math is now using `Decimal()` objects (python's built-in arbitrary precision bignum library) for exact math and rounding everywhere
- uses changes to `ib_async` for more reliable missing bid/ask data detection
- more reliable generation of futures options quotes and ordering of futures options contracts
- passes `mypy` type checks now (not quite --strict yet, but much improved overall)
- replaced ad-hoc delayed event and coroutine scheduler with more well structured [`BGTask`](https://github.com/mattsta/mutil/blob/main/mutil/bgtask.py)
- log files are now placed in `dir/YEAR/MONTH/filename` structure for easier dropping of old logs
- improved/fixed contract cache for stability and accuracy (mainly fixed inside `ib_async` because it was reporting unexpected results for some contract lookups)
- added ability to fetch equity options chains from tradier in 500ms (if you provide an API key) instead of using the multi-minute-slowdown IBKR interface for bulk lookups.
- replaced `pendulum` since it has been unmaintained for years
- even more details and correctness in `buy preview` output for margin impact and credit management
- added a custom runtime predicate language ([`ifthen`](https://github.com/mattsta/tradeapis/blob/main/tradeapis/ifthen.py) over at tradeapis) where you can specify any data field compared against any other data field then a command to run when the condition becomes true.
  - condition checks are performed at _every_ ticker update, and IBKR updates tickers at a maximum rate of 250 ms (240 times per minute) and a minimum rate of "just when something happens"
  - `if AAPL bid <= APPL low: buy AAPL 100 LMT` — if AAPL hits the LOD, just buy 100 shares immediately at the current price
  - symbols can also be "factored out" like `if AAPL { bid <= ask }: say AAPL price at AAPL.bid`
  - the fields available are: high, low, open, close, last, mid/midpoint/live, atr, vwap, iv, delta, theta, gamma
    - it also supports position syntax, so if you have an option quote at row `:23` you could have: `if :23 delta >= 0.50: say delta going up`
  - you can also do arbitrary math: `if AAPL ask >= AAPL high - 0.75: say AAPL near HOD again`
  - there's also a custom percent math mode: `if :23 { bid >= low + 5% }: say row 23 is 5% higher than the low again`
  - or even weird things like: `if AAPL last >= (AAPL high + AAPL low) / 2: say AAPL is over half way between high and low`
  - we also have custom EMA selectors maintained for multiple durations for each live quote instrument (only populated as long as the quote has been running though) for: ema:price, ema:delta, ema:iv
    - `if :23 { ema:delta 300 > ema:delta 900 }: say 5 minute delta rising over 15 minute delta`
  - the predicate `ifthen` system can also consume an external websocket data feed if algo details are being generated somewhere else, so _at runtime_ you can configure which websocket data details can trigger local actions instead of needing to update code for trigger actions.
- quote logic is better abstracted into self-contained `ITicker` objects for holding per-instrument price history stats, ema stats, atr stats, connections between option spreads and legs for calculating spread greeks more easily, and some alert details for when prices are sweeping through ranges
- corrected display digit length in quote toolbar for all instruments using new API-backed instrument DB for correct values everywhere
- much more detailed `info` output for all quote instruments (additional live stats and some derived interpretations of the live stats in addition to just contract details)
- `clear today` for removing all quotes expiring today or older
- `clear expired` for removing all quotes older than today
- `clear unused` for removing all single leg option quotes not used by spread quotes
- `clear options` for removing all options quotes
- `align` command for generating a family of spread quotes at once.
  - `align /ES` by default gives you: ATM strangle, 10 point wide straddle, call spread +10 ATM 20 points wide, put spread -10 ATM 20 points wide.
  - The ATM point offset and width offset can be adjusted as parameters.
- commands can also now have default values for multiple optional parameters
- more IBKR algos enabled: TRAIL LIMIT, SNAP MID/MKT/PRIMARY

**Comment by AlphaSierraHotel** [comment]
Please consider adding  `icli/__pycache__/` to `.gitignore`.

**Reply by mattsta** [reply]
Thanks for the note! I tend to keep those globally and it works well:

```haskell
$ cat ~/.gitignore_global 
.DS_Store
.DS_Store?
._*
.ruff_cache
.Spotlight-V100
.Trashes/
.pytype/
tags
*.dSYM
.rebar/
cscope.out
.python-version
__pycache__/
.pytest.cache
```

**Reply by AlphaSierraHotel** [reply]
That works nicely.  Thanks.

Will you be looking to add the ability to create conditional orders sometime?  I'd like to be able to say, buy these long calls (with a snap-mid) when the price of the underlying reaches a specific price.

**Reply by mattsta** [reply]
In the big description up there, one unreleased feature allows client-side triggers off any field to run any action where you could do things like:

```haskell
if AAPL bid >= 222: buy AAPL241220C00230000 3 LIM
```

and the `buy` command automation does "walking" by default where it submits an order near the current bid, waits 2 seconds for a fill, if not filled, modifies the price higher, waits 2 seconds, and continues until filled or it just hits the ask.

Of course, those triggers are _client_ side and only work as long as your current client is open and receiving live price updates, so they aren't using the more complicated [IBKR conditional order APIs](https://interactivebrokers.github.io/tws-api/order_conditions.html) (but those are also much more confusing and less flexible too).

The `buy` command is also price-aware (since it uses the bid/ask for walking) so you can submit dollar amounts and it figures out quantity itself like `buy AAPL241220C00230000 $2000 LIM` would give you how ever may contracts $2,000 can acquire at the order entry time.

I do need to finally wrap all these new features up into a release, but I'm currently blocked by not having enough funding to justify the time to get it all done consistently. Sponsorships or contracts welcome 🫠

**Reply by AlphaSierraHotel** [reply]
Sweet!  Looking forward to the next release, whenever it comes.

I'm not in a position to contribute anything meaningful at the moment, but I like the idea and hope to one day be able to support the project financially.

**Comment by cdubbm** [comment]
did the above comments make it into the branch as yet?

**Reply by mattsta** [reply]
Not yet, but stay tuned. The main blocker is this update requires publishing updates to 2-3 _other_ projects these new features built on top of, and the more steps causes more delays.

I'm going to throw some code robots at the release process for the other projects and this project soon, so hopefully soon there will be new global capability sync showing up.

i guess like and subscribe for updates?
